### PR TITLE
Pass blocks to component

### PR DESCRIPTION
**Pull request by MikeRogers0**
What do you think of this as a feature (/is there a better way)? The idea is I can add the code:

```
<!-- app/components/panel/_panel.html.erb -->
<div class="panel">
  <div class="panel-body">
    <% yield if block_given? %>
  </div>
</div>
```

```
<!-- Some fancy view -->
<%= render_component("panel") do %>
  <p>Inner content of panel</p>
  <%= render_component("button") %>
<% end %>
```

To my views so I can nest components within components.

If you like the idea, I'll tidy it up & write tests around it and figure out a better way to handle the stubs.

### CODE REVIEW

**Comment by kitop** [comment]
I do like the idea of being able to nest components! @devnacho thoughts?

Implementing stubs might be a bit tricky. But I'm definitively interested in researching this a bit more. Thanks @MikeRogers0!!

**Comment by devnacho** [comment]
@MikeRogers0 I like the idea too, Would be interesting to see it!

**Comment by MikeRogers0** [comment]
Oh it's passing y'all!

Right now it sets the nested content in the local `block` (yield'ing within that view kept being between rails versions). So it can be used like:

```
<!-- app/components/panel/_panel.html.erb -->
<div class="panel">
  <div class="panel-body">
    <%= block %>
  </div>
</div>
```

What do you think? Or would setting it as a property be better (So it can be set as YAML).

File: lib/mountain_view/presenter.rb [hunk 1]
```diff
@@ -10,10 +10,10 @@ def initialize(slug, properties = {})
       @properties = default_properties.deep_merge(properties)
     end
 
-    def render(context)
+    def render(context, &block)
       context.extend ViewContext
       context.inject_component_context self
-      context.render partial: partial
+      context.render partial, partial: partial, block: yield
```
Comment (kitop) [inline lib/mountain_view/presenter.rb]: Curious about passing the partial name twice, `render partial, partial: partial`, what was the reason for that?

**Comment by kitop** [comment]
@MikeRogers0 this looks great! I like the idea of passing the content as a block rather than passing a property as it results in an easy way to nest components.

I'm a bit concerned about the stubs though. How can we show that a component receives a block? And how can we stub the block?
The workaround that you have to do for Rails 4.2 can be an opportunity to be able to stub components by adding a `yield:` property. Although, that would require a bit more work and check in the component, just for rendering the styleguide.

Thoughts?

**Comment by MikeRogers0** [comment]
@kitop Thanks buddy!

> I'm a bit concerned about the stubs though. How can we show that a component receives a block? And how can we stub the block?

I like the idea of passing a `yield:` property & then letting developers pass a suitable stub string to denote that it can be passed a block.

**Comment by kitop** [comment]
@MikeRogers0 interesting! That sounds like a good idea, or at least an starting point.

What do you think about detecting the `yield` property of the stubs and adding a block hint (like `do ... end`) to the styleguide?
https://github.com/devnacho/mountain_view/blob/c293e6081660a76ba2d1639a9f948cc8c194f8f1/app/views/mountain_view/styleguide/show.html.erb#L20-L22

Also, if that yield is detected we can pass the results as the `block` local, or we could somehow make it backreference another stub, like rails fixtures.

I also just remembered that there's a `meta` key in the stubs that we can use to render extra info, so the user can specify more data about the block.

Thoughts?

**Comment by MikeRogers0** [comment]
> What do you think about detecting the yield property of the stubs and adding a block hint (like `do ... end`) to the styleguide?

I have mixed feelings. In my mind if someone has defined they `yield` property they either have another use in mind or they understand its usage. Also I can't find a suitable approach which isn't ugly to read back.

**Comment by MikeRogers0** [comment]
@kitop What do you think so far?

**Comment by kitop** [comment]
@MikeRogers0 sorry, I've been super busy this past couple weeks!
I promise to give it a review very soon! Stay tuned!

**Comment by kitop** [comment]
@MikeRogers0 sorry for the delay, finally have some time

This is looking really good to me!! I think it's good to merge and have a first go at it.

What are your thoughts?
Anything you're feeling concerned about? Or ready to merge too?

**Comment by MikeRogers0** [comment]
I've been using the branch in a new project for the last month, so far no problems! Lets ship!

**Comment by kitop** [comment]
Thanks @MikeRogers0!!

**Comment by MikeRogers0** [comment]
Huzzah awesome!

**Comment by kitop** [comment]
Just released in v0.11.0 🎉

**Pull request merged**